lib/core/src/restore.rs

```diff
@@ -0,0 +1,820 @@
+        // self.send_swap_state_handler
+        //     .update_swap_info(&swap_id, Failed, None, None, Some(&refund_tx_id_str))
+        //     .await?;
+
+        Ok(())
+    }
+    pub(crate) async fn get_onchain_data(
+        &self,
+        tx_map: TxMap,
+        immutable_db: SwapsList,
+    ) -> Result<RecoveredOnchainData> {
+        let histories = self.fetch_swaps_histories(immutable_db).await?;
+
+        let recovered_send_data = self
+            .recover_send_swap_tx_ids(&tx_map, histories.send)
+            .await?;
+        let recovered_receive_data = self
+            .recover_receive_swap_tx_ids(&tx_map, histories.receive)
+            .await?;
+        let recovered_chain_send_data = self
+            .recover_send_chain_swap_tx_ids(&tx_map, histories.send_chain)
+            .await?;
+        let recovered_chain_receive_data = self
+            .recover_receive_chain_swap_tx_ids(&tx_map, histories.receive_chain)
+            .await?;
+
+        Ok(RecoveredOnchainData {
+            send: recovered_send_data,
+            receive: recovered_receive_data,
+            chain_send: recovered_chain_send_data,
+            chain_receive: recovered_chain_receive_data,
+        })
+    }
+
+    /// Reconstruct Send Swap tx IDs from the onchain data and the immutable DB data
+    async fn recover_send_swap_tx_ids(
+        &self,
+        tx_map: &TxMap,
+        send_histories_by_swap_id: HashMap<String, SendSwapHistory>,
+    ) -> Result<RecoveredOnchainDataSend> {
+        let mut lockup_tx_map: HashMap<String, Txid> = HashMap::new();
+        let mut refund_tx_map: HashMap<String, Txid> = HashMap::new();
+
+        for (swap_id, history) in send_histories_by_swap_id {
+            // If a history tx is one of our outgoing txs, it's a lockup tx
+            let maybe_lockup_tx_id = history
+                .iter()
+                .find(|tx_id| tx_map.outgoing_tx_map.contains_key::<Txid>(tx_id));
+            match maybe_lockup_tx_id {
+                None => {
+                    error!("No lockup tx found when recovering data for Send Swap {swap_id}")
+                }
+                Some(lockup_tx_id) => {
+                    lockup_tx_map.insert(swap_id.clone(), *lockup_tx_id);
+                }
+            }
+
+            // If a history tx is one of our incoming txs, it's a refund tx
+            let maybe_refund_tx_id = history
+                .iter()
+                .find(|tx_id| tx_map.incoming_tx_map.contains_key::<Txid>(tx_id));
+            if let Some(refund_tx_id) = maybe_refund_tx_id {
+                refund_tx_map.insert(swap_id.clone(), *refund_tx_id);
+            }
+        }
+
+        info!(
+            "[Recover Send] Found {} lockup txs from onchain data",
+            lockup_tx_map.len()
+        );
+        info!(
+            "[Recover Send] Found {} refund txs from onchain data",
+            refund_tx_map.len()
+        );
+
+        Ok(RecoveredOnchainDataSend {
```

Comment:
As discussed, it makes more sense to have the preimage in the immutable data. It can only be retrieved for some states (non-cooperative claim tx), in other cases its only known from interactions with Boltz. So for consistency, it makes more sense to store it in the immutable data, when available (e.g. after claim is sucessful).